app2/src/lib/utils/convert-display.ts

```diff
@@ -0,0 +1,137 @@
+import { bech32 } from "@scure/base"
+import { Schema } from "effect"
+import { AddressAptosDisplay, AddressCosmosDisplay, AddressEvmDisplay } from "$lib/schema/address"
+
+/**
+ * Convert a bech32 display address to canonical bytes
+ */
+export function cosmosDisplayToCanonical(displayAddress: string): Uint8Array {
+  try {
+    const decoded = bech32.decode(displayAddress as `${string}1${string}`)
+    const canonicalAddress = bech32.fromWords(decoded.words)
+    return new Uint8Array(canonicalAddress)
+  } catch (error: any) {
+    throw new Error(`Invalid Cosmos bech32 address: ${error.message}`)
+  }
+}
+
+/**
+ * Convert an EVM display address (hex) to canonical bytes
+ */
+export function evmDisplayToCanonical(displayAddress: string): Uint8Array {
+  // Validate EVM address format (0x + 40 hex characters)
+  if (!/^0x[0-9a-fA-F]{40}$/.test(displayAddress)) {
+    throw new Error("EVM address must be 0x followed by 40 hex characters")
+  }
+
+  // Remove 0x prefix and convert to bytes
+  const hexWithoutPrefix = displayAddress.slice(2)
+  const bytes = new Uint8Array(20)
+
+  for (let i = 0; i < 40; i += 2) {
+    bytes[i / 2] = Number.parseInt(hexWithoutPrefix.substring(i, i + 2), 16)
+  }
+
+  return bytes
+}
+
+/**
+ * Convert an Aptos display address (hex) to canonical bytes
+ */
+export function aptosDisplayToCanonical(displayAddress: string): Uint8Array {
+  // Validate Aptos address format (0x + 64 hex characters)
+  if (!/^0x[0-9a-fA-F]{64}$/.test(displayAddress)) {
+    throw new Error("Aptos address must be 0x followed by 64 hex characters")
+  }
+
+  // Remove 0x prefix and convert to bytes
```

Comment:
for these functions, make the displayAddress inputs the appropiate displayaddress types: https://github.com/unionlabs/union/blob/main/app2/src/lib/schema/address.ts

and the return type AddressCanonicalBytes. This prevents misuse (ie you put a cosmos display address into the aptos function). Once that is done, don't throw an error but return an `Option<T>`. It will in practice never be `None` unless we mess something up in our translation logic, because we have already ensured that the incoming displayAddress has the correct pattern for each of the variants as those regexes are encoded in the schema